enhancements/ingress/optional-ingress-hypershift.md

```diff
@@ -0,0 +1,293 @@
+The cluster ingress operator becomes unnecessary and can be disabled to:
+- reduce the resource consumption on
+  - the HyperShift client clusters (ingress controllers)
+  - the HyperShift hosted control plane (operator)
+- make the HyperShift hosted control plane lighter and simpler, therefore more manageable
+- enable the HyperShift cluster users to make their choice for the cluster ingress
+
+### Goals
+- Implement the new ingress capability.
+- Add the ingress capability to the ingress operator's payload.
+- Make the dependent cluster operators tolerate the absence of the ingress operator.
+- Do not lock the subsequent implementation for the standalone OpenShift.
+
+### Non-Goals
+- Use a different way for enabling/disabling the ingress other than via [the cluster capability](/enhancements/installer/component-selection.md).
+- Fully implement the ingress capability for the standalone OpenShift:
+  - Make the OpenShift installer tolerate the absence of the default ingress controller.
+  - Make the cluster authentication operator tolerate the absence of the default ingress controller.
+  - Make the cluster monitoring operator tolerate the absence of the default ingress controller.
+- Disable the Route API or try to tolerate the absence of the default ingress controller.
+- Design the capability API for HyperShift.
+- Implement the ingress operator disabling in the HyperShift's control plane operator.
+
+## Proposal
+
+### User Stories
+
+- As a HyperShift engineer, I want to be able to disable the ingress operator on the clusters where the ingress is implemented differently so that the hosted control plane can become lighter.
+- As a HyperShift engineer, I want the cluster operators to tolerate the absence of the ingress operator to keep the clusters from being degraded.
+- As a HyperShift hosted cluster's user, I want to see the ingress capability as a known capability to bring the visibility to all the cluster actors (admins, users, other operators).
+- As a HyperShift hosted cluster's user, I want the default ingress controller to not occupy the cluster resources while the ingress traffic is routed by the cloud provider's load balancers.
+
+### Workflow Description
+
+The HyperShift hosted cluster's user wants to use an AWS ALB to route the traffic to its example application deployed on a HyperShift's hosted cluster.
+
+#### Terminology
+[HyperShift concepts and personas](https://hypershift-docs.netlify.app/reference/concepts-and-personas/).
+
+#### Disable the default ingress
+
+- Cluster service consumer requests a hosted cluster without the ingress capability using the HostedCluster API (**note**: the HostedCluster's capability API does not exist and it's out of scope of this enhancement).
+- HyperShift creates the hosted cluster.
+- HyperShift skips the deployment of the ingress operator on the hosted control plane.
+- HyperShift creates `ClusterVersion` without the ingress capability on the hosted cluster.
```

Comment:
This likely implies the creation of a new Capability Set that includes everything the default/all-caps set includes today, minus Ingress.

It's either that or hypershift needs to use the "None" capset, and then explicitly list all the caps except ingress, and update that list as new caps are added to the product.  (This is slightly nicer from a user perspective since it doesn't expose a capset definition to standalone OCP users, when it can't be used there, but also much less preferable from a maintenance/risk of keeping it updated perspective)

Not a serious obstacle, but worth calling out as explicit implementation requirements so we know how we're going to do it.

Reply:
I agree, that phrase is vague, mostly due to the lack of the vision on the future HyperShift API. I'm fine to put a new implementation detail about the creation of a dedicated "baseline sets without ingress" if the HyperShift engineering finds it necessary. However I thought of it from the conservative perceptive:  currently the only way to disable a cluster capability is by using `None` baseline - let HyperShift adapt to it. Also, a new "capability set without ingress" may be more convenient for HyperShift but more puzzling for the standalone users. I can envision a scenario where users decide to try out a new 'baseline set without the ingress' and, to their surprise, find out that the ingress was added by CVO it the end.

@csrwng: Do you have any idea of how you would like a cluster capability to be disabled? Note that currently the cluster version's API doesn't allow to exclude capabilities from the baselines. So if you the HyperShift's capability API will look exactly like cluster verison's one - the user will have to use `None` baseline and add any needed capablities except `Ingress`.

Reply:
Hi @alebedev87 it sounds fine to not introduce a new capability set for this. As you mentioned, we can handle making it optional through some other means in the HyperShift API for now.

Reply:
@csrwng it seems like we have a reasonable approach now in which we introduce the capability, but it can't be disabled on standalone clusters(for now).  Is that ok w/ you?